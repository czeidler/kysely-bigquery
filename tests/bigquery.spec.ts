import { expect, test } from 'vitest'

import { BigQueryDialect } from '../src'
import { Kysely } from 'kysely';

const kysely = new Kysely<any>({
  dialect: new BigQueryDialect()
})

test('simple select', async () => {
  const query = kysely.selectFrom('features.metadata').where('id', '>', 10).selectAll().limit(1);

  expect(query.compile()).toMatchInlineSnapshot(`
    {
      "parameters": [
        10,
        1,
      ],
      "query": {
        "from": {
          "froms": [
            {
              "kind": "TableNode",
              "table": {
                "identifier": {
                  "kind": "IdentifierNode",
                  "name": "metadata",
                },
                "kind": "SchemableIdentifierNode",
                "schema": {
                  "kind": "IdentifierNode",
                  "name": "features",
                },
              },
            },
          ],
          "kind": "FromNode",
        },
        "kind": "SelectQueryNode",
        "limit": {
          "kind": "LimitNode",
          "limit": {
            "kind": "ValueNode",
            "value": 1,
          },
        },
        "selections": [
          {
            "kind": "SelectionNode",
            "selection": {
              "kind": "SelectAllNode",
            },
          },
        ],
        "where": {
          "kind": "WhereNode",
          "where": {
            "kind": "BinaryOperationNode",
            "leftOperand": {
              "column": {
                "column": {
                  "kind": "IdentifierNode",
                  "name": "id",
                },
                "kind": "ColumnNode",
              },
              "kind": "ReferenceNode",
              "table": undefined,
            },
            "operator": {
              "kind": "OperatorNode",
              "operator": ">",
            },
            "rightOperand": {
              "kind": "ValueNode",
              "value": 10,
            },
          },
        },
      },
      "sql": "select * from \`features\`.\`metadata\` where \`id\` > ? limit ?",
    }
  `)

  const row = await query.execute();

  expect(row).toMatchInlineSnapshot(`
    [
      {
        "agg_universe": "n.a.",
        "category": "cac",
        "compounding": "n.a.",
        "created_at": BigQueryTimestamp {
          "value": "2022-06-24T06:28:56.000Z",
        },
        "description": "Monthly customer acquisition cost (CAC). Marketing costs only.",
        "id": 28,
        "inserted_at": BigQueryTimestamp {
          "value": "2023-11-14T00:01:53.000Z",
        },
        "interval_measured_quantity": 1,
        "interval_measured_units": "month",
        "name": "cac",
        "name_raw": "cac_marketing",
        "public": 0,
        "subcategory": "amount",
        "type": "feature",
        "updated_at": BigQueryTimestamp {
          "value": "2023-11-13T23:10:54.000Z",
        },
      },
    ]
  `)

  
})
