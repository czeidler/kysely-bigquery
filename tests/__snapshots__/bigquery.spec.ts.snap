// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`introspection 1`] = `
[
  {
    "columns": [
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "bank_account_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "remote_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "fpath_account_id",
      },
      {
        "dataType": "DATE",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "date",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "original_description",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "pending",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "processing_status",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "amount",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "iso_currency_code",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "unofficial_currency_code",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "bank_account_category_id",
      },
      {
        "dataType": "DATE",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "authorized_date",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "name",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "merchant_name",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "pending_transaction_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "transaction_type",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "account_owner",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "amount_usd",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "transaction_category",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "transaction_status",
      },
      {
        "dataType": "STRUCT<type STRING, name STRING>",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "counter_party",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "created_at",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "updated_at",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "__hevo_xmin",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "__hevo__database_name",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "__hevo__schema_name",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "__hevo__ingested_at",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "__hevo__loaded_at",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "ntropy_batch_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enriched_by",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_merchant_name",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat3",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat2",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat4",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat1",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enriched_at",
      },
      {
        "dataType": "STRUCT<recurrence_group STRUCT<total_amount FLOAT64, average_amount FLOAT64, average_amount_per_tx FLOAT64, id STRING, other_party STRING, periodicity STRING, date_of_last_tx STRING, transaction_ids STRING, date_of_first_tx STRING, frequency_in_days FLOAT64, first_payment_date STRING, latest_payment_date STRING, periodicity_in_days FLOAT64>, mcc STRING, logo STRING, labels STRING, website STRING, merchant STRING, merchant_id STRING, transaction_id STRING, location STRING, confidence FLOAT64, recurrence STRING, person STRING, intermediaries STRING>",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment",
      },
    ],
    "isView": false,
    "name": "bank_account_transactions",
    "schema": "hevo",
  },
  {
    "columns": [
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "bank_account_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "remote_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "fpath_account_id",
      },
      {
        "dataType": "DATE",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "date",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "original_description",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "pending",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "processing_status",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "amount",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "iso_currency_code",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "unofficial_currency_code",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "bank_account_category_id",
      },
      {
        "dataType": "DATE",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "authorized_date",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "name",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "merchant_name",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "pending_transaction_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "transaction_type",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "account_owner",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "amount_usd",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "transaction_category",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "transaction_status",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "counter_party",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "created_at",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "updated_at",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "ntropy_batch_id",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enriched_at",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enriched_by",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat1",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat2",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat3",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat4",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_merchant_name",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "inserted_at",
      },
    ],
    "isView": false,
    "name": "bank_account_transactions",
    "schema": "api",
  },
]
`;
